src/Castle.Core/DynamicProxy/Generators/MetaMethod.cs

```diff
@@ -101,7 +102,30 @@ internal override void SwitchToExplicitImplementation()
 				Attributes |= MethodAttributes.SpecialName;
 			}
 
-			name = string.Format("{0}.{1}", Method.DeclaringType.Name, Method.Name);
+			var nameBuilder = new StringBuilder();
+			AppendTypeNameTo(nameBuilder, Method.DeclaringType);
+			nameBuilder.Append('.');
+			nameBuilder.Append(Method.Name);
+			name = nameBuilder.ToString();
+		}
+
+		private static void AppendTypeNameTo(StringBuilder stringBuilder, Type type)
+		{
+			stringBuilder.Append(type.Name);
+			if (type.GetTypeInfo().IsGenericType)
+			{
+				stringBuilder.Append('[');
+				var genericTypeArguments = type.GetGenericArguments();
+				for (int i = 0, n = genericTypeArguments.Length; i < n; ++i)
```

Comment:
I don't think the performance will be affected. This only gets run when `SwitchToExplicitImplementation` is called once per `MetaMethod` and it is then cached in its `name` field. The `StringBuilder` will perform pretty similar to the existing `string.Format` and users are rarely going to have more than a single level of generic nesting, and even then the cost here is negligible compared to the actual type generation or other work DP does.